In [2]:
import numpy as np
from scipy import stats
import statsmodels.api as sm

# ТЕСТ 1: ОДНОВЫБОРОЧНЫЙ Z-ТЕСТ ДЛЯ СРЕДНЕЙ (σ известна)

**СЦЕНАРИЙ:**
Компания производит батарейки. Заявленное среднее время работы = 500 часов.
Из партии взято 50 батареек. Известно σ = 20 часов.
Проверяем, соответствует ли выборка заявленному среднему значению.

**ГИПОТЕЗЫ:**
H₀: μ = 500 часов (батарейки работают как заявлено)
H₁: μ ≠ 500 часов (двусторонний тест)

In [4]:
n1 = 50
sigma1 = 20
mu0_1 = 500

# Генерируем данные
np.random.seed(42)
sample1 = np.random.normal(loc=505, scale=sigma1, size=n1)

# Вычисляем выборочную среднюю
x_mean1 = np.mean(sample1)

print(f"Данные:")
print(f"Размер выборки (n): {n1}")
print(f"Выборочная средняя (x̄): {x_mean1:.2f} часов")
print(f"Известное σ: {sigma1} часов")
print(f"Гипотетическое μ_0: {mu0_1} часов")

# Z-статистика
z_stat1 = (x_mean1 - mu0_1) / (sigma1 / np.sqrt(n1))

# P-value (двусторонний тест)
p_value1 = 2 * (1 - stats.norm.cdf(np.abs(z_stat1)))

# Критическое значение (α = 0.05)
alpha = 0.05
z_critical1 = stats.norm.ppf(1 - alpha/2)

print(f"Z-статистика: z = {z_stat1:.4f}")
print(f"P-value (двусторонний): {p_value1:.4f}")
print(f"Критическое значение (z_0.975): ±{z_critical1:.4f}")

Данные:
Размер выборки (n): 50
Выборочная средняя (x̄): 500.49 часов
Известное σ: 20 часов
Гипотетическое μ_0: 500 часов
Z-статистика: z = 0.1734
P-value (двусторонний): 0.8623
Критическое значение (z_0.975): ±1.9600


**Интерпретация**
P-value = 0.8623, что больше чем α = 0.05. Следовательно, гипотезу H₀ не отклоняем. 
Также z = 0.1734 < 1.9600, что также говорит о корректности гипетезы H₀. 
Таким образом среднее время работы не отличается статистически значимо от 500 часов.

# ТЕСТ 2: ОДНОВЫБОРОЧНЫЙ t-ТЕСТ ДЛЯ СРЕДНЕЙ (σ неизвестна)

**СЦЕНАРИЙ:**
Магазин одежды проверяет среднее время обслуживания клиента.
Нормальное время = 15 минут. Из 30 клиентов измерили время обслуживания.
σ неизвестна, поэтому используем t-тест.

**ГИПОТЕЗЫ:**
H₀: μ = 15 минут
H₁: μ ≠ 15 минут

In [5]:
n2 = 30
mu0_2 = 15
# Генерируем данные
sample2 = np.random.normal(loc=14.5, scale=2.5, size=n2)
# Выборочная статистика
x_mean2 = np.mean(sample2)
s2 = np.std(sample2, ddof=1)
print(f"ДАННЫЕ:")
print(f"Размер выборки (n): {n2}")
print(f"Выборочная средняя (x̄): {x_mean2:.2f} минут")
print(f"Выборочное стандартное отклонение (s): {s2:.2f} минут")
print(f"Гипотетическое μ_0: {mu0_2} минут")
# t-статистика
t_stat2 = (x_mean2 - mu0_2) / (s2 / np.sqrt(n2))
df2 = n2 - 1
p_value2 = 2 * (1 - stats.t.cdf(np.abs(t_stat2), df2))
t_critical2 = stats.t.ppf(1 - alpha/2, df2)

print(f"t-статистика: t = {t_stat2:.4f}")
print(f"Степени свободы (df): {df2}")
print(f"P-value (двусторонний): {p_value2:.4f}")
print(f"Критическое значение (t_29*0.975): ±{t_critical2:.4f}")

ДАННЫЕ:
Размер выборки (n): 30
Выборочная средняя (x̄): 14.61 минут
Выборочное стандартное отклонение (s): 2.47 минут
Гипотетическое μ_0: 15 минут
t-статистика: t = -0.8559
Степени свободы (df): 29
P-value (двусторонний): 0.3991
Критическое значение (t_29*0.975): ±2.0452


**Интерпретация**
P-value = 0.3991, что больше чем α = 0.05. Следовательно, гипотезу H₀ не отклоняем. 
Также t = -0.8559 > -2.0452, что также говорит о корректности гипетезы H₀.
Таким образом среднее время обслуживания не отличается статистически значимо от 15 минут.

# ТЕСТ 3: ОДНОВЫБОРОЧНЫЙ Z-ТЕСТ ДЛЯ ДОЛИ

**СЦЕНАРИЙ:**
Интернет-компания заявляет, что доля довольных клиентов = 0.80 (80%).
Из 200 клиентов опросили 165 довольных.
Проверяем, соответствует ли выборка заявленной доле.

**ГИПОТЕЗЫ:**
H₀: p = 0.80 (доля довольных = 80%)
H₁: p ≠ 0.80

In [6]:
n3 = 200
x3 = 165
p0_3 = 0.80
# Выборочная доля
p_hat3 = x3 / n3

print(f"ДАННЫЕ:")
print(f"Размер выборки (n): {n3}")
print(f"Количество довольных (x): {x3}")
print(f"Выборочная доля (p̂): {p_hat3:.4f} ({p_hat3*100:.1f}%)")
print(f"Гипотетическая доля (p_0): {p0_3:.4f} ({p0_3*100:.1f}%)")

# Z-тест для доли
se3 = np.sqrt(p0_3 * (1 - p0_3) / n3)
z_stat3 = (p_hat3 - p0_3) / se3
p_value3 = 2 * (1 - stats.norm.cdf(np.abs(z_stat3)))
z_critical3 = stats.norm.ppf(1 - alpha/2)

print(f"Z-статистика: z = (p̂ - p_0) / SE = {z_stat3:.4f}")
print(f"P-value: {p_value3:.4f}")

ДАННЫЕ:
Размер выборки (n): 200
Количество довольных (x): 165
Выборочная доля (p̂): 0.8250 (82.5%)
Гипотетическая доля (p_0): 0.8000 (80.0%)
Z-статистика: z = (p̂ - p_0) / SE = 0.8839
P-value: 0.3768


**Интерпретация**
P-value = 0.3768, что больше чем α = 0.05. Следовательно, гипотезу H₀ не отклоняем. 
Таким образом довольных не отличается статистически значимо от 80%.

# ТЕСТ 4: ДВУХВЫБОРОЧНЫЙ Z-ТЕСТ ДЛЯ ДОЛЕЙ

**СЦЕНАРИЙ:**
Две компании конкурируют. Проверяем, различаются ли доли клиентов,
использующих их приложения.
- Компания A: из 300 опросов, 180 используют
- Компания B: из 250 опросов, 140 используют

**ГИПОТЕЗЫ:**
H₀: p₁ = p₂ (доли пользователей одинаковые)
H₁: p₁ ≠ p₂ (доли различаются)

In [16]:
n1_prop, x1_prop = 300, 180
n2_prop, x2_prop = 250, 140

p_hat1_prop = x1_prop / n1_prop
p_hat2_prop = x2_prop / n2_prop
p_pool = (x1_prop + x2_prop) / (n1_prop + n2_prop)

print(f"ДАННЫЕ:")
print(f"Компания A: n_1={n1_prop}, x_1={x1_prop}, p̂_1={p_hat1_prop:.4f}")
print(f"Компания B: n_2={n2_prop}, x_2={x2_prop}, p̂_2={p_hat2_prop:.4f}")
print(f"Объединённая доля: p_pool = {p_pool:.4f}")

# Z-тест для двух долей
se_prop = np.sqrt(p_pool * (1 - p_pool) * (1/n1_prop + 1/n2_prop))
z_stat4 = (p_hat1_prop - p_hat2_prop) / se_prop
p_value4 = 2 * (1 - stats.norm.cdf(np.abs(z_stat4)))
z_critical4 = stats.norm.ppf(1 - alpha/2)

print(f"Z-статистика: z = {z_stat4:.4f}")
print(f"P-value: {p_value4:.4f}")
se = np.sqrt(p_pool * (1 - p_pool) * (1/n1 + 1/n2))
print(f"SE: {se:.4f}")

ДАННЫЕ:
Компания A: n_1=300, x_1=180, p̂_1=0.6000
Компания B: n_2=250, x_2=140, p̂_2=0.5600
Объединённая доля: p_pool = 0.5818
Z-статистика: z = 0.9470
P-value: 0.3437
SE: 0.1139


**Интерпретация**
P-value = 0.3437, что больше чем α = 0.05. Следовательно, гипотезу H₀ не отклоняем. 
Доли пользователей не различаются статистически значимо.

# ТЕСТ 5: ДВУХВЫБОРОЧНЫЙ Z-ТЕСТ ДЛЯ СРЕДНЕЙ (σ известны)

**СЦЕНАРИЙ:**
Сравниваем зарплаты в двух филиалах компании.
- Филиал A: σ₁ = 5000, n₁ = 60
- Филиал B: σ₂ = 4500, n₂ = 55

In [18]:
n1_mean, sigma1_mean = 60, 5000
n2_mean, sigma2_mean = 55, 4500

# Генерируем данные
sample1_mean = np.random.normal(loc=50000, scale=sigma1_mean, size=n1_mean)
sample2_mean = np.random.normal(loc=48000, scale=sigma2_mean, size=n2_mean)

x_mean1_5 = np.mean(sample1_mean)
x_mean2_5 = np.mean(sample2_mean)

print(f"ДАННЫЕ:")
print(f"Филиал A: x̄_1 = {x_mean1_5:.2f}, σ_1 = {sigma1_mean}, n_1 = {n1_mean}")
print(f"Филиал B: x̄_2 = {x_mean2_5:.2f}, σ_2 = {sigma2_mean}, n_2 = {n2_mean}")

# Z-тест для двух средних
se_mean = np.sqrt(sigma1_mean**2 / n1_mean + sigma2_mean**2 / n2_mean)
z_stat5 = (x_mean1_5 - x_mean2_5) / se_mean
p_value5 = 2 * (1 - stats.norm.cdf(np.abs(z_stat5)))

print(f"Z-статистика: z = {z_stat5:.4f}")
print(f"P-value: {p_value5:.4f}")

ДАННЫЕ:
Филиал A: x̄_1 = 49339.52, σ_1 = 5000, n_1 = 60
Филиал B: x̄_2 = 47592.44, σ_2 = 4500, n_2 = 55
Z-статистика: z = 1.9721
P-value: 0.0486


**Интерпретация**
P-value = 0.0486, что меньше чем α = 0.05. Следовательно, различия в средних статистически значимо

# ТЕСТ 6: ДВУХВЫБОРОЧНЫЙ t-ТЕСТ С РАВНЫМИ ДИСПЕРСИЯМИ (POOLED)

**СЦЕНАРИЙ:**
Маркетинговое агентство сравнивает эффективность двух стратегий продаж.
Стратегия A: 25 дней
Стратегия B: 22 дня
Предположение: σ_1 = σ_2

In [12]:
# Генерируем данные
n1_test6, n2_test6 = 25, 22
sample1_test6 = np.random.normal(loc=12, scale=2.5, size=n1_test6)
sample2_test6 = np.random.normal(loc=10.5, scale=2.5, size=n2_test6)

# Выборочная статистика
x_mean1_6 = np.mean(sample1_test6)
x_mean2_6 = np.mean(sample2_test6)
s1_test6 = np.std(sample1_test6, ddof=1)
s2_test6 = np.std(sample2_test6, ddof=1)

# Объединённая дисперсия
sp2 = ((n1_test6 - 1) * s1_test6**2 + (n2_test6 - 1) * s2_test6**2) / (n1_test6 + n2_test6 - 2)
sp = np.sqrt(sp2)

print(f"ДАННЫЕ:")
print(f"Стратегия A: x̄_1 = {x_mean1_6:.2f}, s_1 = {s1_test6:.2f}, n_1 = {n1_test6}")
print(f"Стратегия B: x̄_2 = {x_mean2_6:.2f}, s_2 = {s2_test6:.2f}, n_2 = {n2_test6}")

# t-тест
df_test6 = n1_test6 + n2_test6 - 2
se_pooled = sp * np.sqrt(1/n1_test6 + 1/n2_test6)
t_stat6 = (x_mean1_6 - x_mean2_6) / se_pooled
p_value6 = 2 * (1 - stats.t.cdf(np.abs(t_stat6), df_test6))
t_critical6 = stats.t.ppf(1 - alpha/2, df_test6)

print(f"Объединённая дисперсия: s_p^2 = {sp2:.2f}")
print(f"t-статистика: t = {t_stat6:.4f}")
print(f"Степени свободы: df = {df_test6}")
print(f"P-value: {p_value6:.4f}")
print(f"Критическое значение: ±{t_critical6:.4f}")

ДАННЫЕ:
Стратегия A: x̄_1 = 12.67, s_1 = 2.22, n_1 = 25
Стратегия B: x̄_2 = 10.36, s_2 = 2.84, n_2 = 22
Объединённая дисперсия: s_p^2 = 6.41
t-статистика: t = 3.1242
Степени свободы: df = 45
P-value: 0.0031
Критическое значение: ±2.0141


**Интерпретация**
P-value = 0, что меньше чем α = 0.05. 
Кроме того,  t = 3.1242 > 2.0141. Следовательно, эффективности стратегий статистически значимо отличается

# ТЕСТ 7: ДВУХВЫБОРОЧНЫЙ t-ТЕСТ С РАЗЛИЧНЫМИ ДИСПЕРСИЯМИ (WELCH'S)

**СЦЕНАРИЙ:**
Компания доставки сравнивает время доставки в два региона.
Регион A: более стабильная доставка (σ_1 меньше)
Регион B: менее стабильная доставка (σ_2 больше)
Предположение: σ_1 ≠ σ_2

In [15]:
# Генерируем данные с разными дисперсиями
n1_test7, n2_test7 = 20, 18
sample1_test7 = np.random.normal(loc=3, scale=0.5, size=n1_test7)
sample2_test7 = np.random.normal(loc=4, scale=1.3, size=n2_test7)

# Выборочная статистика
x_mean1_7 = np.mean(sample1_test7)
x_mean2_7 = np.mean(sample2_test7)
s1_test7 = np.std(sample1_test7, ddof=1)
s2_test7 = np.std(sample2_test7, ddof=1)

print(f"ДАННЫЕ:")
print(f"Регион A: x̄_1 = {x_mean1_7:.2f}, s_1 = {s1_test7:.2f}, n_1 = {n1_test7}")
print(f"Регион B: x̄_2 = {x_mean2_7:.2f}, s_2 = {s2_test7:.2f}, n_2 = {n2_test7}")

# Welch's t-тест
se_welch = np.sqrt(s1_test7**2 / n1_test7 + s2_test7**2 / n2_test7)
t_stat7 = (x_mean1_7 - x_mean2_7) / se_welch

# Степени свободы (Welch-Satterthwaite)
numerator_df = (s1_test7**2/n1_test7 + s2_test7**2/n2_test7)**2
denominator_df = ((s1_test7**2/n1_test7)**2/(n1_test7-1) + 
                  (s2_test7**2/n2_test7)**2/(n2_test7-1))
df_welch = numerator_df / denominator_df

p_value7 = 2 * (1 - stats.t.cdf(np.abs(t_stat7), df_welch))
t_critical7 = stats.t.ppf(1 - alpha/2, df_welch)

print(f"t-статистика (Welch): t = {t_stat7:.4f}")
print(f"Степени свободы: df ≈ {df_welch:.2f}")
print(f"P-value: {p_value7:.4f}")
print(f"Критическое значение: ±{t_critical7:.4f}")

ДАННЫЕ:
Регион A: x̄_1 = 2.89, s_1 = 0.51, n_1 = 20
Регион B: x̄_2 = 3.81, s_2 = 1.09, n_2 = 18
t-статистика (Welch): t = -3.3003
Степени свободы: df ≈ 23.52
P-value: 0.0031
Критическое значение: ±2.0661


**Интерпретация**
P-value = 0.0031, что меньше чем α = 0.05. 
Кроме того,  t = -3.3003 < -2.0661. 
Следовательно, среднее время доставки различается статистически значимо.

# Заключение

Были созданы бизнес сценарии и сгенерированы данные к ним для проведения 7 различных тестов:
1.  Одновыборочный Z-тест для Средней ($\sigma$ известна)
2.  Одновыборочный t-тест для Средней ($\sigma$ неизвестна)
3.  Одновыборочный Z-тест для Доли (Пропорции)
4.  Двухвыборочный Z-тест для Долей (Пропорций)
5.  Двухвыборочный Z-тест для Средней ($\sigma$ известны)
6.  Двухвыборочный t-тест с равными дисперсиями (Pooled)
7.  Двухвыборочный t-тест с различными дисперсиями (Welch’s)

По результатам каждого из тестов была дана интерпретация полученного результата.